### PR TITLE
Fixing MySQL In Docker Unit Tests

### PR DESCRIPTION
**Pull request by nniehoff**
# Closes: #1516 

# What's Changed

Changing the MySQL recommended collation setting from `utf8mb4_bin` to `utf8mb4_general_ci` in order for unit tests to pass inside of a docker container.

# TODO

- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/docs/installation/migrating-from-postgresql.md [hunk 1]
```diff
@@ -18,7 +18,7 @@ This will result in a file named `nautobot_dump.json`.
 
 ## Create the MySQL database
 
-Create the MySQL database for Nautobot, ensuring that it has the correct character set (`utf8mb4`) and collation (`utf8mb4_bin`) settings for case-sensitivity (MySQL will otherwise be case-insensitive by default, which may cause problems when importing your data from the case-sensitive PostgreSQL database dump). Refer to the instructions for [CentOS/RHEL](./centos.md) or [Ubuntu](./ubuntu.md) as necessary if you are unsure how to set up MySQL and create the Nautobot database.
+Create the MySQL database for Nautobot, ensuring that it has the correct character set (`utf8mb4`) and collation (`utf8mb4_general_ci`) settings for case-sensitivity (MySQL will otherwise be case-insensitive by default, which may cause problems when importing your data from the case-sensitive PostgreSQL database dump). Refer to the instructions for [CentOS/RHEL](./centos.md) or [Ubuntu](./ubuntu.md) as necessary if you are unsure how to set up MySQL and create the Nautobot database.
```
Comment (glennmatthews) [inline nautobot/docs/installation/migrating-from-postgresql.md:21]: This is the one point that gives me pause here - in a Postgres deployment, the database itself is case-sensitive (e.g. you can create devices "DEVICE1" and "device1" and they will coexist just fine) but case-insensitive queries are possible. It seems that in MySQL it's all or nothing - either the database and its queries are case-sensitive, or the database and its queries are case-insensitive. So I'm no longer certain that we should across-the-board recommend one or the other collation setting with MySQL - we should probably document the pros and cons of both options clearly and leave it up to the user to decide.

Comment (jathanism) [reply nautobot/docs/installation/migrating-from-postgresql.md:21]: Agreed. This sketches me out here. Why are these tests only failing in Docker containers, but not otherwise? I would like to better understand this.

Comment (jathanism) [reply nautobot/docs/installation/migrating-from-postgresql.md:21]: These docs will need to be revised to instead say that MySQL is case-insensitive by default, and because of this you will need to account for this in your database dumps PRIOR to migrating from PostgreSQL to MySQL. We may also consider offering some ways to determine this ahead of time?

Comment (bryanculver) [reply nautobot/docs/installation/migrating-from-postgresql.md:21]: After a deep excursion into the depths of Django's lookup strategy in Postgres v MySQL, I do not think it's recommended to advise people create case sensitive MySQL databases. Django doesn't expect it at all and it's `i*` expressions are not built to handle it.

TLDR: For Postgres Django does essentially `UPPER(field) = UPPER(lookup)` for iexact. However for MySQL does `field LIKE lookup`. This is insufficient when MySQL is running a case sensitive collation. This seems to be a nastier tradeoff than cleaning up data on migrate. If you need case sensitivity, you should use Postgres.

**Review by bryanculver — CHANGES_REQUESTED** [review]
Just needs docs updates.

Nice find @jathanism regarding the Django lookup differences between Postgres and MySQL. Looks like it's non-trivial to run MySQL with case sensitivity.

**Review by bryanculver — COMMENTED** [review]
Just needs docs updates.

Nice find @jathanism regarding the Django lookup differences between Postgres and MySQL. Looks like it's non-trivial to run MySQL with case sensitivity.

File: nautobot/docs/installation/centos.md [hunk 2]
```diff
@@ -156,7 +156,7 @@ owners.
 
 Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
 
-mysql> CREATE DATABASE nautobot CHARACTER SET utf8mb4 COLLATE utf8mb4_bin;
+mysql> CREATE DATABASE nautobot;
```
Comment (glennmatthews) [inline nautobot/docs/installation/centos.md:161]: Is this removed because mysql defaults to the correct character set and collation if you don't specify anything? Have we confirmed this on centos specifically?

File: development/docker-compose.mysql.yml [hunk 3]
```diff
@@ -15,7 +15,6 @@ services:
       - "NAUTOBOT_DB_ENGINE=django.db.backends.mysql"
   db:
     image: mysql:8
-    command: "--character-set-server=utf8mb4 --collation-server=utf8mb4_bin"
```
Comment (glennmatthews) [inline development/docker-compose.mysql.yml:18]: Is this removed because mysql 8 already defaults to the correct server-side character set and collation?

File: nautobot/docs/installation/migrating-from-postgresql.md [hunk 4]
```diff
@@ -8,17 +8,39 @@ In your existing installation of Nautobot with PostgreSQL, run the following com
...
 ```no-highlight
 (nautobot-postgres) $ nautobot-server dumpdata \
-    --natural-foreign --natural-primary \
-    --exclude contenttypes --exclude auth.permission --exclude extras.job \
-    --format json --indent 2 --traceback \
+    --natural-foreign \
+    --natural-primary \
+    --exclude contenttypes \
+    --exclude auth.permission \
+    --exclude extras.job \
```
Comment (glennmatthews) [inline nautobot/docs/installation/migrating-from-postgresql.md:15]: Probably should remove this line in 1.3, actually...

File: nautobot/docs/installation/ubuntu.md [hunk 5]
```diff
@@ -112,7 +112,7 @@ owners.
 
 Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
 
-mysql> CREATE DATABASE nautobot CHARACTER SET utf8mb4 COLLATE utf8mb4_bin;
+mysql> CREATE DATABASE nautobot;
```
Comment (glennmatthews) [inline nautobot/docs/installation/ubuntu.md:115]: Same question here - has this been checked on ubuntu specifically?

File: nautobot/docs/installation/upgrading.md [hunk 6]
```diff
@@ -25,6 +25,27 @@ Please see the section on [migrating to Celery from RQ](./services.md#migrating-
...
+
+The collation encoding is used to inform MySQL how characters are sorted in the database. This is important when it comes to retrieving data that has special characters or special byte-encoding such as accents or ligatures, and also including emojis. In some cases, with the `utf8mb4_bin` encoding we were previously recommending, case-insensitive searching may return inconsistent or incorrect results.
+
+!!! danger
+	It is **strongly recommended** that you backup your database before exuecitng this query and that you perform this in a non-production environment to identify any potential issues prior to updating your production environment.
+
+If you have an existing MySQL database, you may update your database to use the recommended encoding by using `nautobot-server dbshell` to launch a database shell and executing the `ALTER DATABASE nautobot COLLATE utf8mb4_0900_ai_ci;`query:
```
Comment (glennmatthews) [inline nautobot/docs/installation/upgrading.md:39]:
```suggestion
If you have an existing MySQL database, you may update your database to use the recommended encoding by using `nautobot-server dbshell` to launch a database shell and executing the following command:
```

File: nautobot/docs/installation/centos.md (hunk 2 above)
Comment (jathanism) [reply nautobot/docs/installation/centos.md:161]: Yes on both counts.

File: development/docker-compose.mysql.yml (hunk 3 above)
Comment (jathanism) [reply development/docker-compose.mysql.yml:18]: Yep!

File: nautobot/docs/installation/migrating-from-postgresql.md (hunk 4 above)
Comment (jathanism) [inline nautobot/docs/installation/migrating-from-postgresql.md:15]:
```suggestion
```

File: nautobot/docs/installation/ubuntu.md (hunk 5 above)
Comment (jathanism) [reply nautobot/docs/installation/ubuntu.md:115]: Yep!

File: nautobot/docs/installation/migrating-from-postgresql.md [hunk 7]
```diff
@@ -8,17 +8,39 @@ In your existing installation of Nautobot with PostgreSQL, run the following com
...
+Create the MySQL database for Nautobot, ensuring that it is utilizing the default character set (`utf8mb4`) and default collation (`utf8mb4_0900_ai_ci`) settings for case-insensitivity. It is required that MySQL will be case-insensitive. Because these encodings are the defaults, if your MySQL installation has not been modified, there will be nothing for you to do other than make sure.
+
+In very rare cases, there may problems when importing your data from the case-sensitive PostgreSQL database dump that will need to be handled on a case-by-case basis. Please refer to the instructions for [CentOS/RHEL](./centos.md) or [Ubuntu](./ubuntu.md) as necessary if you are unsure how to set up MySQL and create the Nautobot database.
+
+### Confirming database encoding
+
+To confirm that your MySQL database has the correct encoding, you may start up a database shell using `nautobot-server dbshell` and run the `SELECT @@character_set_database, @@collation_database;` query:
```
Comment (jathanism) [inline nautobot/docs/installation/migrating-from-postgresql.md:32]:
```suggestion
To confirm that your MySQL database has the correct encoding, you may start up a database shell using `nautobot-server dbshell` and run the following command:
```

**Review by glennmatthews — APPROVED** [review]

**Review by bryanculver — APPROVED** [review]
